rad/chain.rad

```diff
@@ -38,52 +55,91 @@
 
-(def enter-chain
-  (fn [eval-definer chain]
-    (def old-eval eval)
-    (def ch (ref chain))
-    (def ev
-      (fn [e]
-        (if (eq? e :quit)
-            (do (print! "leaving chain")
-                (eval-definer old-eval))
-          (do (def x (eval-in-chain e (read-ref ch)))
-              (write-ref ch (view (@ :chain) x))
-              (view (@ :result) x)))))
-    (eval-definer ev)))
-
-(document 'enter-chain
-          '(("eval-definer" function) ("chain" chain))
-          "Instantiates an 'eval' which evaluates expressions in 'chain'. The expression ':quit' will return to the previous eval. Uses the parameter 'eval-definer' for eval redefinition, this should be a function which can be used to redefine 'eval': it should set eval to it's input.")
-
-(def enter-remote-chain
-  (fn [eval-definer chain-id]
-    (def old-eval eval)
-    (def chain (update-chain-with empty-chain chain-id (fn [e] '()) (fn [e] '())))
-    (print! (string-append "Synced with remote chain, at index: " (show (view (@ :index) chain))))
-    (def ch (ref chain))
-    (def inputs (ref (list)))
-    (eval-definer
-     (fn [e]
-       (cond
-        (eq? e :quit) (do (print! "Leaving remote-chain.")
-                          (eval-definer old-eval))
-        (eq? e :send) (do (print! "Sending inputs to remote chain.")
-                          (map (fn [i] (send! chain-id i)) (reverse (read-ref inputs)))
-                          (print! "Done! Leaving remote-chain.")
-                          (eval-definer old-eval))
-        :else (do (def x (eval-in-chain e (read-ref ch)))
-                  (write-ref ch (view (@ :chain) x))
-                  (modify-ref inputs (fn [is] (cons e is)))
-                  (view (@ :result) x)))))))
-
-(document 'enter-remote-chain
-          '(("eval-definer" function) ("chain-id" string))
-          "Syncs up to a remote chain, and allows experimentation inputs. There are two special inputs: ':quit' will leave the chain, ':send' will send all the accumulated inputs to the remote chain, one by one, and then leave the chain. Uses the parameter 'eval-definer' for eval redefinition, this should be a function which can be used to redefine 'eval': it should set eval to it's input.")
-
-;; Setup an eval definer
-(def boom
-  (quote
-   (do (def eval-ref (ref eval))
-       (def eval-definer (fn [f] (write-ref eval-ref f)))
-       (def eval (fn [e] ((read-ref eval-ref) e))))))
+;; update-chain
+
+(def update-chain
+  (fn [chain]
+    (def next (ref chain))
+    (update-chain-with chain (lookup :name chain) (fn [x] (modify-ref next (fn [r] (eval-in-chain x r)))) print!)))
+
+(document 'update-chain
+		  '(("chain" chain))
+		  "Return a version of 'chain' with new inputs and correspondingly updated state")
+
+;; add-quit
+
+(def add-quit
+  (fn [old-state new-eval]
```

Comment:
Feels to me that the argument `new-eval` is more like `old-eval` (this is an eval-transformer function?).

Reply:
Yeah, I think both 'old-state' and 'new-eval' are confusing names. I used them because I think of *entering* a new eval with the `:quit`, and then `:quit`-ing is going *back* to the old state. But I'll rename (and add some docs).